src/Network/Transport/TCP/Internal.hs

```diff
@@ -137,3 +144,14 @@ recvExact sock len = go [] len
       if BS.null bs
         then throwIO (userError "recvExact: Socket closed")
         else go (bs : acc) (l - fromIntegral (BS.length bs))
+
+data LockingChan a = LC !(MVar ()) !(Chan a)
+
+newLockingChan :: IO (LockingChan a)
+newLockingChan = LC <$> newMVar () <*> newChan
+
+withLockingChan :: LockingChan a -> (Chan a -> IO b) -> IO b
+withLockingChan (LC lock chan) f = withMVar lock (const $ f chan)
+
+unsafeWithLockingChan :: LockingChan a -> (Chan a -> IO b) -> IO b
+unsafeWithLockingChan (LC _ ch) f = f ch
```

Comment:
Functions need documentation, with a clear statement of the invariant they enforce. Otherwise, it's unclear what's unsafe about the "unsafe" variants.